apiserver/pkg/storage/calico/blockAffinity_storage.go

```diff
@@ -0,0 +1,129 @@
+		return nil, fmt.Errorf("Unable to update block affinity. Block affinity resources are read-only.")
+	}
+	getFn := func(ctx context.Context, c clientv3.Interface, ns string, name string, opts clientOpts) (resourceObject, error) {
+		ogo := opts.(options.GetOptions)
+		return c.BlockAffinities().Get(ctx, name, ogo)
+	}
+	deleteFn := func(ctx context.Context, c clientv3.Interface, ns string, name string, opts clientOpts) (resourceObject, error) {
+		return nil, fmt.Errorf("Unable to delete block affinity. Block affinity resources are read-only.")
+	}
+	listFn := func(ctx context.Context, c clientv3.Interface, opts clientOpts) (resourceListObject, error) {
+		olo := opts.(options.ListOptions)
+		return c.BlockAffinities().List(ctx, olo)
+	}
+	watchFn := func(ctx context.Context, c clientv3.Interface, opts clientOpts) (watch.Interface, error) {
+		olo := opts.(options.ListOptions)
+		return c.BlockAffinities().Watch(ctx, olo)
+	}
+	dryRunnableStorage := registry.DryRunnableStorage{Storage: &resourceStore{
+		client:            c,
+		codec:             opts.RESTOptions.StorageConfig.Codec,
+		versioner:         etcd.APIObjectVersioner{},
+		aapiType:          reflect.TypeOf(aapi.BlockAffinity{}),
+		aapiListType:      reflect.TypeOf(aapi.BlockAffinityList{}),
+		libCalicoType:     reflect.TypeOf(libapi.BlockAffinity{}),
+		libCalicoListType: reflect.TypeOf(libapi.BlockAffinityList{}),
+		isNamespaced:      false,
+		create:            createFn,
+		update:            updateFn,
+		get:               getFn,
+		delete:            deleteFn,
+		list:              listFn,
+		watch:             watchFn,
+		resourceName:      "BlockAffinity",
+		converter:         BlockAffinityConverter{},
+	}, Codec: opts.RESTOptions.StorageConfig.Codec}
+	return dryRunnableStorage, func() {}
+}
+
+type BlockAffinityConverter struct {
+}
+
+func (gc BlockAffinityConverter) convertToLibcalico(aapiObj runtime.Object) resourceObject {
+	aapiBlockAffinity := aapiObj.(*aapi.BlockAffinity)
+	lcgBlockAffinity := &libapi.BlockAffinity{}
+	lcgBlockAffinity.TypeMeta = aapiBlockAffinity.TypeMeta
+	lcgBlockAffinity.ObjectMeta = aapiBlockAffinity.ObjectMeta
+	lcgBlockAffinity.Kind = libapi.KindBlockAffinity
+	lcgBlockAffinity.APIVersion = aapi.GroupVersionCurrent
+	lcgBlockAffinity.Spec.State = aapiBlockAffinity.Spec.State
+	lcgBlockAffinity.Spec.Node = aapiBlockAffinity.Spec.Node
+	lcgBlockAffinity.Spec.CIDR = aapiBlockAffinity.Spec.CIDR
+
+	// Deleted is an internal field and should be set to false for any block affinity picked up by the AAPI.
+	lcgBlockAffinity.Spec.Deleted = false
+
+	return lcgBlockAffinity
+}
+
+func (gc BlockAffinityConverter) convertToAAPI(libcalicoObject resourceObject, aapiObj runtime.Object) {
+	lcgBlockAffinity := libcalicoObject.(*libapi.BlockAffinity)
+	aapiBlockAffinity := aapiObj.(*aapi.BlockAffinity)
+	// Copy spec but ignore internal field Deleted.
+	aapiBlockAffinity.Spec.State = lcgBlockAffinity.Spec.State
```

Comment:
The more I think about it, the more I think we might want to share the same struct here to avoid the need to copy fields individually. Will be a huge pain when we forget to do this when we add a new field.

I think we can do it in a Part II PR, though.